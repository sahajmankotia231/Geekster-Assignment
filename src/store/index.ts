import { createSlice, configureStore } from "@reduxjs/toolkit";

export type ChildType = {
  age: number | null;
};

export type RoomType = {
  maxNumberOfAdults: number;
  maxNumberOfChildren: number;
  maxOccupancy: number;
  adultsCounter: number;
  totalGuests: number;
  children: ChildType[];
};

export type StoreStateType = {
  maxNumberOfRooms: number;
  rooms: RoomType[];
};

const initialRoomsState = {
  maxNumberOfAdults: 5,
  maxNumberOfChildren: 3,
  maxOccupancy: 5,
  adultsCounter: 1,
  totalGuests: 1,
  children: [],
};

const initialState = {
  maxNumberOfRooms: 8,
  rooms: [initialRoomsState],
};

const counterSlice = createSlice({
  name: "rooms",
  initialState,
  reducers: {
    initializeStore: (state, action): void => {
      if (Boolean(action.payload)) {
        const splitedRooms = action.payload?.split("|") || [];
        const rooms = splitedRooms?.map(
          (room): RoomType => {
            const guests = room.split(":");
            const adultsCounter = Number(guests[0]);
            const children =
              guests[1]?.split(",")?.reduce((arr, age): ChildType[] => {
                arr.push({ age: Number(age) });
                return arr;
              }, []) || [];

            return {
              ...initialRoomsState,
              adultsCounter,
              children,
            };
          }
        );

        state.rooms = rooms;
      }
    },
    resetStore: (state): void => {
      state.rooms = [initialRoomsState];
    },
    addRoom: (state): void => {
      const { rooms, maxNumberOfRooms } = state;

      if (rooms.length < maxNumberOfRooms) {
        rooms.push(initialRoomsState);
      }
    },
    removeRoom: (state, action): void => {
      if (state.rooms.length > 1) {
        state.rooms.splice(action.payload, 1);
      }
    },
    addAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.adultsCounter < room.maxNumberOfAdults &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.adultsCounter += 1;
        room.totalGuests += 1;
      }
    },
    removeAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (room.adultsCounter > 1) {
        room.adultsCounter -= 1;
        room.totalGuests -= 1;
      }
    },
    addChild: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.children.length < room.maxNumberOfChildren &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.children.push({
          age: null,
        });
        room.totalGuests += 1;
      }
    },
    removeChild: (state, action): void => {
      const room = state.rooms[action.payload];
      const { children } = room;

      if (children.length) {
        children.splice(children.length - 1, 1);
        room.totalGuests -= 1;
      }
    },
    removeSpecificChild: (state, action): void => {
      const { childrenPosition, roomPosition } = action.payload;
      const room = state.rooms[roomPosition];
      const { children } = room;

      if (children.length) {
        children?.splice(childrenPosition, 1);
        room.totalGuests -= 1;
      }
    },
    changeChildAge: (state, action): void => {
      const { childrenPosition, roomPosition, age } = action.payload;
      const { children } = state.rooms[roomPosition];

      if (Boolean(age)) {
        children[childrenPosition].age = Number(age);
      }
    },
  },
});

export const {
  initializeStore,
  resetStore,
  addRoom,
  removeRoom,
  addAdult,
  removeAdult,
  addChild,
  removeChild,
  removeSpecificChild,
  changeChildAge,
} = counterSlice.actions;

const store = configureStore({
  reducer: counterSlice.reducer,
});

export default store;
