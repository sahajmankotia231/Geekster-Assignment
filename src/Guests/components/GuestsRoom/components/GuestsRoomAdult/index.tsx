import React, { ReactElement } from "react";
import { useSelector, useDispatch } from "react-redux";

import Button from "components/Button";
import { ReactComponent as IconPlus } from "assets/plus.svg";
import { ReactComponent as IconMinus } from "assets/minus.svg";
import { StoreStateType, RoomType, addAdult, removeAdult } from "store";

import { Guest, GuestText, GuestButtonGroup } from "../../styles";

type GuestsRoomAdultsProps = {
  position: number;
};

const GuestsRoomAdult = ({ position }: GuestsRoomAdultsProps): ReactElement => {
  const dispatch = useDispatch();
  const {
    adultsCounter,
    maxNumberOfAdults,
    totalGuests,
    maxOccupancy,
  }: RoomType = useSelector((state: StoreStateType) => state.rooms[position]);
  const isRemoveButtonDisabled = adultsCounter === 1;
  const isAddButtonDisabled =
    adultsCounter === maxNumberOfAdults || maxOccupancy === totalGuests;

  return (
    <Guest>
      <GuestText>Adults</GuestText>
      <GuestButtonGroup>
        <Button
          onClick={() => dispatch(removeAdult(position))}
          theme="secondary"
          dataTestId={`remove-adult-${position}-button`}
          disabled={isRemoveButtonDisabled}
        >
          <IconMinus />
        </Button>
        <GuestText data-testid="adults-counter">{adultsCounter}</GuestText>
        <Button
          onClick={() => dispatch(addAdult(position))}
          theme="secondary"
          dataTestId={`add-adult-${position}-button`}
          disabled={isAddButtonDisabled}
        >
          <IconPlus />
        </Button>
      </GuestButtonGroup>
    </Guest>
  );
};

export default GuestsRoomAdult;
