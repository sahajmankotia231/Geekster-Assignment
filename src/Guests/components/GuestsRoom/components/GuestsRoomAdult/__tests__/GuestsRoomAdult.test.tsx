import React from "react";
import { Provider } from "react-redux";
import { render, fireEvent, screen } from "@testing-library/react";

import store, { resetStore } from "store";

import GuestsRoomAdult from "..";

describe("<GuestsRoomAdult />", () => {
  beforeEach(() => store.dispatch(resetStore()));

  it("should render properly", () => {
    const { container } = render(
      <Provider store={store}>
        <GuestsRoomAdult position={0} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should be able to increase and decrease the number of adults", () => {
    render(
      <Provider store={store}>
        <GuestsRoomAdult position={0} />
      </Provider>
    );

    const adultsCounter = screen.getByTestId("adults-counter");

    expect(adultsCounter.textContent).toBe("1");

    fireEvent.click(screen.getByTestId("add-adult-0-button"));
    expect(adultsCounter.textContent).toBe("2");

    fireEvent.click(screen.getByTestId("remove-adult-0-button"));
    expect(adultsCounter.textContent).toBe("1");
  });

  it("should disable remove button if the room have just one guest", () => {
    render(
      <Provider store={store}>
        <GuestsRoomAdult position={0} />
      </Provider>
    );

    expect(screen.getByTestId("remove-adult-0-button")).toHaveAttribute(
      "disabled"
    );
  });

  it("should disable add button in the room max of guests", () => {
    const room = store.getState().rooms[0];

    render(
      <Provider store={store}>
        <GuestsRoomAdult position={0} />
      </Provider>
    );

    const addButton = screen.getByTestId("add-adult-0-button");

    expect(addButton).not.toHaveAttribute("disabled");

    for (let i = 1; i <= room.maxNumberOfAdults + 1; i++) {
      fireEvent.click(addButton);
    }

    expect(addButton).toHaveAttribute("disabled");
  });
});
