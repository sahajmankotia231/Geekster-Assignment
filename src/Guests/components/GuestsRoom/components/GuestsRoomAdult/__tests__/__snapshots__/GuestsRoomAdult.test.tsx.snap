// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GuestsRoomAdult /> should render properly 1`] = `
<div
  class="sc-eCImvq guCNbL"
>
  <span
    class="sc-jRQAMF bgpgfc"
  >
    Adults
  </span>
  <div
    class="sc-gKckTs jduWvm"
  >
    <button
      class="sc-bdvvaa fzcnyp"
      data-testid="remove-adult-0-button"
      disabled=""
    >
      <svg>
        minus.svg
      </svg>
    </button>
    <span
      class="sc-jRQAMF bgpgfc"
      data-testid="adults-counter"
    >
      1
    </span>
    <button
      class="sc-bdvvaa fzcnyp"
      data-testid="add-adult-0-button"
    >
      <svg>
        plus.svg
      </svg>
    </button>
  </div>
</div>
`;
